lib/evaluate/fold.cc

```diff
@@ -152,35 +157,486 @@ ComplexPart FoldOperation(FoldingContext &context, ComplexPart &&complexPart) {
       FoldOperation(context, std::move(complex)), complexPart.part()};
 }
 
+// Scalar conversion
+// Conversion Helper to be used to avoid instantiating ScalarConvert when
+// it would be undefined.
+template<TypeCategory TOCAT, TypeCategory FROMCAT>
+static constexpr inline bool IsAllowedBasicConversion() {
+  if constexpr (TOCAT == TypeCategory::Integer || TOCAT == TypeCategory::Real) {
```

Comment:
Thanks for pointing to it. I could not find a good name for this predicate, the goal here is to tell whether a `Convert<TO, FROMCAT>` is allowed with these type categories. From the comment in the `Convert` class in `lib/evaluate/expression.h` I think COMPLEX conversions should not be allowed here so I am not sure I can use `common::IsNumericTypeCategory`.
However, the predicate here is the same as the one inside the static_assertion in the Convert class. Do you think it is worth finding a better name and moving this to a common header so that it can also be used in the Convert class to avoid duplication?

Reply:
Which is better?

`if (TOCAT == TypeCategory::Integer || TOCAT == TypeCategory::Real) { ... `

or

`if (common::TypeCategoryIsIntegerOrReal(TOCAT) { ... `

I think that the first one is better.  The function doesn't provide an abstraction, doesn't make anything more readable, and adds a needless level of indirection to the reader of the code.

Reply:
I see your point. For now I have remove all the implicit conversions to "bigger" types when folding with host runtime in order to focus the new PR #292 on the host runtime indirection feature. If we want such implicit conversions again, I will keep this in mind.